internal/tracer/span_collector.go

```diff
@@ -0,0 +1,120 @@
+package tracer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	exporttrace "go.opentelemetry.io/otel/sdk/export/trace"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// SpanCollector does 3 things:
+// 1) Accepts spans from OpenTelemetry.
+// 2) Stores spans (for now, in memory)
+// 3) Allows consumers to read spans they might want to send elsewhere
+// Numbers 2 and 3 access the same data, and so it's a concurrency issue.
+type SpanCollector struct {
+
+	// members for communicating with the loop() goroutine
+
+	// for OpenTelemetry SpanCollector
+	spanDataCh chan *exporttrace.SpanData
+
+	// for SpanSource
+	readReqCh chan chan []*exporttrace.SpanData
+	requeueCh chan []*exporttrace.SpanData
+}
+
+// SpanSource is the interface for consumers (generally telemetry.Controller)
+type SpanSource interface {
+	// GetOutgoingSpans gives a consumer access to spans they should send
+	// If there are no outoing spans, err will be io.EOF
+	// rejectFn allows client to reject spans, so they can be requeued
+	GetOutgoingSpans() (data io.Reader, rejectFn func(), err error)
+}
+
+func NewSpanCollector(ctx context.Context) *SpanCollector {
+	r := &SpanCollector{
+		spanDataCh: make(chan *exporttrace.SpanData),
+		readReqCh:  make(chan chan []*exporttrace.SpanData),
+		requeueCh:  make(chan []*exporttrace.SpanData),
+	}
+	go r.loop(ctx)
+	return r
+}
+
+func (c *SpanCollector) loop(ctx context.Context) {
+	// spans that have come in and are waiting to be read by a consumer
+	var queue []*exporttrace.SpanData
+
+	for {
```

Comment:
Should this goroutine ever stop? Maybe when Shutdown is called?

Reply:
Just added it; lmk if you want to chat through it.

Reply:
LGTM!